### PR TITLE
two hop flattener transformer

### PR DESCRIPTION
**Pull request by Juliamg**

### CODE REVIEW

File: tests/tests_unit/graph/test_transformers/test_prune_graph.py [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+        (namespace["Text-Destination-ID"], namespace["value"], Literal("Vacuum system self-check completed.")),
+        (namespace["Text-Destination-ID"], namespace["description"], Literal("vacuum status")),
+        (namespace["Device-Source-ID"], namespace["textProperty"], namespace["Text-Destination-ID"]),
+        (namespace["Device-Source-ID"], RDF_TYPE, namespace["Device"]),
+    ]
+
+    expected_triples_new_predicate = [
+        (namespace["Device-Source-ID"], RDF_TYPE, namespace["Device"]),
+        (namespace["Device-Source-ID"], namespace["vacuumStatus"], Literal("Vacuum system self-check completed.")),
+    ]
```
Comment (doctrino) [inline tests/tests_unit/graph/test_transformers/test_prune_graph.py:58]: I like this, makes it very nice to understand what is going to happen in the transformer.

**Review by doctrino — APPROVED** [review]
This is solid work. I think we don't add anything to the changelog for this one, as it is an internal transformer.

File: cognite/neat/_graph/transformers/_prune_graph.py [hunk 2]
```diff
@@ -1,84 +1,131 @@
...
-                                     ?destinationNode <{property_predicate}> ?{property_name} . }}"""
+
+    _query_template_keep_old_predicate: str = """
+    SELECT ?sourceNode ?property_source ?destinationNode ?property_destination ?property_value WHERE {{
+        ?sourceNode ?property_source ?destinationNode .
+        ?destinationNode ?property_destination ?property_value .
+        ?destinationNode a <{destination_node_type}> .
+        ?destinationNode <{value_property_predicate}> ?property_value . }}"""
+
+    _query_template_new_predicate: str = """
```
Comment (nikokaoja) [inline cognite/neat/_graph/transformers/_prune_graph.py:65]: more easier to read query (IMO):

```


"""SELECT ?sourceNode ?sourceProperty ?destinationNode ?newSourceProperty ?newSourcePropertyValue WHERE {{
          ?destinationNode a <{destination_node_type}> .
          ?destinationNode <{destinationProperty}> ?newSourcePropertyValue .
          ?destinationNode <{destinationPropertyHoldingNewPropertyName}> ?newSourceProperty .
          ?sourceNode ?sourceProperty ?destinationNode . }}"""


```

File: cognite/neat/_graph/transformers/_prune_graph.py [hunk 3]
```diff
@@ -1,84 +1,131 @@
...
+        ?destinationNode <{predicate_property_predicate}> ?predicate_value .
+        ?destinationNode <{value_property_predicate}> ?property_value . }}"""
 
     def __init__(
         self,
         destination_node_type: URIRef,
-        property_predicate: Namespace,
-        property_name: str,
-        delete_connecting_node: bool = True,
+        namespace: Namespace,
```
Comment (nikokaoja) [inline cognite/neat/_graph/transformers/_prune_graph.py:75]: namespace of what? I guess of destinationProperty which holds value you are interested in?

if so , why not calling it that way? `destination_property_namespace` ?

File: cognite/neat/_graph/transformers/_prune_graph.py [hunk 4]
```diff
@@ -1,84 +1,131 @@
...
+        value_property_name: str,
```
Comment (nikokaoja) [inline cognite/neat/_graph/transformers/_prune_graph.py:76]: rename to `destination_property_name`

File: cognite/neat/_graph/transformers/_prune_graph.py [hunk 5]
```diff
@@ -1,84 +1,131 @@
...
+        delete_connecting_node: bool = False,
+        predicate_property_name: str | None = None,
```
Comment (nikokaoja) [inline cognite/neat/_graph/transformers/_prune_graph.py:78]: rename to `property_name_holding_new_property` maybe ? 

lift this arg above `delete_connecting_node`

File: cognite/neat/_graph/transformers/_prune_graph.py [hunk 6]
```diff
@@ -1,84 +1,131 @@
...
+            )
+        else:
+            query = self._query_template_keep_old_predicate.format(
+                destination_node_type=self.destination_node_type,
+                value_property_predicate=self.value_property_predicate,
+            )
+
+        graph_traversals = graph.query(query)
+
+        for result_raw in graph_traversals:
```
Comment (nikokaoja) [inline cognite/neat/_graph/transformers/_prune_graph.py:104]:
```suggestion
        for (source_node, predicate, destination_node, property_value,) in graph.query(query):
```

you can already here unpack results

File: cognite/neat/_graph/transformers/_prune_graph.py [hunk 7]
```diff
@@ -1,84 +1,131 @@
...
+            source_node, old_predicate, destination_node, new_predicate_value, new_property_value = (
+                result_raw.asdict().values()
             )
```
Comment (nikokaoja) [inline cognite/neat/_graph/transformers/_prune_graph.py:107]:
```suggestion
```

File: cognite/neat/_graph/transformers/_prune_graph.py [hunk 8]
```diff
@@ -1,84 +1,131 @@
...
-
-        graph_traversals = list(
-            graph.query(
-                self._query_template.format(
-                    destination_node_type=self.destination_node_type,
-                    property_predicate=self.property_predicate,
-                    property_name=self.property_name,
-                )
+    def transform(self, graph) -> None:
+        nodes_to_delete: list = []
```
Comment (nikokaoja) [inline cognite/neat/_graph/transformers/_prune_graph.py:87]:
```suggestion
        nodes_to_delete: list[tuple] = []
```

File: cognite/neat/_graph/transformers/_prune_graph.py [hunk 9]
```diff
@@ -1,84 +1,131 @@
...
+                predicate = as_neat_compliant_uri(self.namespace[new_predicate_value_string])
+            else:
+                predicate = old_predicate
+
+            # Create new connection from source node to value
+            graph.add((source_node, predicate, new_property_value))
+            # Remove old relationship between source node and destination node
+            graph.remove((source_node, predicate, destination_node))
+
+            nodes_to_delete.append(destination_node)
```
Comment (nikokaoja) [inline cognite/neat/_graph/transformers/_prune_graph.py:121]: you need to store both source and destination node
```suggestion
            nodes_to_delete.append((source_node, destination_node))
```

File: cognite/neat/_graph/transformers/_prune_graph.py [hunk 10]
```diff
@@ -1,84 +1,131 @@
...
+
+        if self.delete_connecting_node:
+            for node in nodes_to_delete:
```
Comment (nikokaoja) [inline cognite/neat/_graph/transformers/_prune_graph.py:124]:
```suggestion
            for (source_node, destination_node,) in nodes_to_delete:
```

File: cognite/neat/_graph/transformers/_prune_graph.py [hunk 11]
```diff
@@ -1,84 +1,131 @@
...
+                # Remove edge triples to node
+                graph.remove((None, None, node))
```
Comment (nikokaoja) [inline cognite/neat/_graph/transformers/_prune_graph.py:126]:
```suggestion
                graph.remove((source_node, None, destination_node))
```

File: cognite/neat/_graph/transformers/_prune_graph.py [hunk 12]
```diff
@@ -1,84 +1,131 @@
...
+                # Remove node triple
+                graph.remove((node, None, None))
```
Comment (nikokaoja) [inline cognite/neat/_graph/transformers/_prune_graph.py:128]:
```suggestion
                graph.remove((destination_node, None, None))
```

**Review by nikokaoja — APPROVED** [review]
Two things, which I think are very personal

1. Names of variables 
2. Ordering of sparql statements

File: cognite/neat/_graph/transformers/_prune_graph.py [hunk 13]
```diff
@@ -1,84 +1,131 @@
...
+        (destination_node_type = ns1:Attribute, namespace = ns1, value_property_name="Value",
+        predicate_property_name = "Name", delete_connectiong_node=True):
 
-        Graph after flattening nodes with destination_node_type = rdf:type(TextObject), property_predicate = :value,
-        and property_name = "value":
+        <http://purl.org/cognite/neat/Pump-1> a ns1:Pump ;
+        ns1:ItemTagAssignmentClass "D-20PSV0002" .
 
-        node(A, rdf:type(Pump)) -(predicate("vendor"))> Literal("CompanyX")
+        Number of triples after operation: 5 - 1*3 = 2
```
Comment (nikokaoja) [inline cognite/neat/_graph/transformers/_prune_graph.py:45]: as we have been using source and target in our code base suggest renaming destination to target

File: cognite/neat/_graph/transformers/_prune_graph.py [hunk 14]
```diff
@@ -1,84 +1,131 @@
...
-from rdflib.term import Identifier
+
+from cognite.neat._utils.rdf_ import as_neat_compliant_uri
+from cognite.neat._utils.text import sentence_or_string_to_camel
 
 from ._base import BaseTransformer
 
 
-# TODO: Handle the cse when value is None, which will not make the TextObject resolve
 class TwoHopFlattener(BaseTransformer):
```
[the remark this answers is not in the document]
Comment (Juliamg) [reply cognite/neat/_graph/transformers/_prune_graph.py:10]: what about `AttachPropertyFromTargetToSource` ? And I will elaborate in docstring

[the remark this answers is not in the document]
Comment (Juliamg) [reply cognite/neat/_graph/transformers/_prune_graph.py:10]: Also, Use Case A will use the old predicate, i.e. `:sourceProperty` and not `:propertyWhichValueWeWant`. This is how the transformation was used for IODD.

File: cognite/neat/_graph/transformers/_prune_graph.py (hunk 3 above)
Comment (Juliamg) [reply cognite/neat/_graph/transformers/_prune_graph.py:75]: It is the namespace for the entire graph actually. Not restricted to destination node

File: cognite/neat/_graph/transformers/_prune_graph.py (hunk 9 above)
Comment (Juliamg) [reply cognite/neat/_graph/transformers/_prune_graph.py:121]: I am already deleting that triple in row 119 though. So I just want to collect the TargetNode to remove it if the `delete_connecting_node` flag is set.

**Pull request merged**